File: frontend/src/erd/prismaIdentifiers.ts
import {
  PRISMA_IDENTIFIER_CONTRACT_VERSION,
  PRISMA_IDENTIFIER_MAX_ATTEMPTS,
  isPrismaIdentifier,
  isReservedPrismaName,
} from "./prismaIdentifierContract";

export type PrismaIdentifierKind = "model" | "field" | "relation";

export type PrismaIdentifierMapping = {
  kind: PrismaIdentifierKind;
  namespace: string;
  source: string;
  generated: string;
};

export type PrismaIdentifierFailure = {
  key: string;
  kind: PrismaIdentifierKind;
  namespace: string;
  source: string;
  preferred: string;
  lastCandidate: string;
  attempts: number;
  maxAttempts: number;
};

export type PrismaIdentifierRequest = {
  key: string;
  kind: PrismaIdentifierKind;
  namespace: string;
  source: string;
  preferred?: string;
};

export type PrismaIdentifierAllocation = {
  ok: boolean;
  names: Map<string, string>;
  mappings: PrismaIdentifierMapping[];
  failure?: PrismaIdentifierFailure;
};

/**
 * Collapse a database or canvas name to a Prisma grammar candidate.
 *
 * Non-ASCII, punctuation, and whitespace become `_`. Leading digits use the
 * `M_` prefix. Reserved names take a trailing `_` so they do not collide with
 * a source already named `M_model`.
 */
export function preferredPrismaName(source: string): string {
  const normalized = source.normalize("NFC");
  let candidate = normalized.replace(/[^A-Za-z0-9_]+/g, "_");
  candidate = candidate.replace(/_+/g, "_").replace(/^_+|_+$/g, "");
  if (candidate.length === 0) {
    candidate = "unnamed";
  }
  if (!/^[A-Za-z]/.test(candidate)) {
    candidate = `M_${candidate}`;
  }
  if (isReservedPrismaName(candidate) || !isPrismaIdentifier(candidate)) {
    candidate = `${candidate}_`;
  }
  return candidate;
}

function compareRequests(
  left: PrismaIdentifierRequest,
  right: PrismaIdentifierRequest,
): number {
  const sourceOrder = compareUnicode(left.source, right.source);
  if (sourceOrder !== 0) {
    return sourceOrder;
  }
  const namespaceOrder = compareUnicode(left.namespace, right.namespace);
  if (namespaceOrder !== 0) {
    return namespaceOrder;
  }
  return compareUnicode(left.key, right.key);
}

function compareUnicode(left: string, right: string): number {
  let leftIndex = 0;
  let rightIndex = 0;
  while (leftIndex < left.length && rightIndex < right.length) {
    const leftCodePoint = left.codePointAt(leftIndex) ?? 0;
    const rightCodePoint = right.codePointAt(rightIndex) ?? 0;
    if (leftCodePoint !== rightCodePoint) {
      return leftCodePoint < rightCodePoint ? -1 : 1;
    }
    leftIndex += leftCodePoint > 0xffff ? 2 : 1;
    rightIndex += rightCodePoint > 0xffff ? 2 : 1;
  }
  if (leftIndex === left.length && rightIndex === right.length) {
    return 0;
  }
  return leftIndex === left.length ? -1 : 1;
}

/**
 * Allocate deterministic, collision-free Prisma identifiers.
 *
 * Requests are sorted by source text so input order cannot change the mapping.
 * A taken or reserved candidate receives `_2`, `_3`, … until the attempt bound.
 */
export function allocatePrismaIdentifiers(
  requests: PrismaIdentifierRequest[],
  maxAttempts: number = PRISMA_IDENTIFIER_MAX_ATTEMPTS,
): PrismaIdentifierAllocation {
  const names = new Map<string, string>();
  const mappings: PrismaIdentifierMapping[] = [];
  const usedByNamespace = new Map<string, Set<string>>();
  const ordered = [...requests].sort(compareRequests);

  for (const request of ordered) {
    const used = usedByNamespace.get(request.namespace) ?? new Set<string>();
    usedByNamespace.set(request.namespace, used);
    const base = request.preferred ?? preferredPrismaName(request.source);
    let candidate = base;
    let suffix = 2;
    let attempts = 0;
    while (
      used.has(candidate) ||
      !isPrismaIdentifier(candidate) ||
      isReservedPrismaName(candidate)
    ) {
      attempts += 1;
      if (attempts > maxAttempts) {
        return {
          ok: false,
          names,
          mappings,
          failure: {
            key: request.key,
            kind: request.kind,
            namespace: request.namespace,
            source: request.source,
            preferred: base,
            lastCandidate: candidate,
            attempts,
            maxAttempts,
          },
        };
      }
      candidate = `${base}_${suffix}`;
      suffix += 1;
    }
    used.add(candidate);
    names.set(request.key, candidate);
    mappings.push({
      kind: request.kind,
      namespace: request.namespace,
      source: request.source,
      generated: candidate,
    });
  }

  return { ok: true, names, mappings };
}

/**
 * Build the export manifest recorded beside a generated schema.
 */
export function buildPrismaManifest(
  mappings: PrismaIdentifierMapping[],
  failure?: PrismaIdentifierFailure,
): {
  contractVersion: string;
  mappings: PrismaIdentifierMapping[];
  failure?: PrismaIdentifierFailure;
} {
  const manifest = {
    contractVersion: PRISMA_IDENTIFIER_CONTRACT_VERSION,
    mappings,
  };
  return failure ? { ...manifest, failure } : manifest;
}

File: backend/app/spec/prisma_identifiers.py
"""Collision-free Prisma identifier allocation for snapshot ORM export.

The grammar and reserved-name tables are pinned to the Prisma Schema API and
the prisma-engines reserved-model-name list. Canvas export uses the same
contract in ``frontend/src/erd/prismaIdentifierContract.ts``.
"""

from __future__ import annotations

import json
import re
import unicodedata
from typing import Any, Literal

PRISMA_IDENTIFIER_PATTERN = re.compile(r"^[A-Za-z][A-Za-z0-9_]*$")
PRISMA_IDENTIFIER_CONTRACT_VERSION = "2026-08-16.prisma-6-reserved"
PRISMA_IDENTIFIER_MAX_ATTEMPTS = 10_000
PRISMA_EXPORT_FAILURE_SCHEMA = (
    "// Prisma export failed.\n"
    "// A unique identifier could not be allocated within the configured bound.\n"
    "// Download the diagnostic manifest, rename colliding tables or columns, then export again.\n"
)

PRISMA_SCHEMA_KEYWORDS = frozenset(
    {"datasource", "enum", "generator", "model", "type", "view"}
)
PRISMA_SCALAR_TYPE_NAMES = frozenset(
    {
        "BigInt",
        "Boolean",
        "Bytes",
        "DateTime",
        "Decimal",
        "Float",
        "Int",
        "Json",
        "String",
        "Unsupported",
    }
)
PRISMA_CLIENT_RESERVED_NAMES = frozenset(
    {
        "PrismaClient",
        "async",
        "await",
        "break",
        "case",
        "catch",
        "class",
        "const",
        "continue",
        "debugger",
        "default",
        "delete",
        "do",
        "else",
        "enum",
        "export",
        "extends",
        "false",
        "finally",
        "for",
        "function",
        "if",
        "implements",
        "import",
        "in",
        "instanceof",
        "interface",
        "let",
        "new",
        "null",
        "package",
        "private",
        "protected",
        "public",
        "return",
        "super",
        "switch",
        "this",
        "throw",
        "true",
        "try",
        "typeof",
        "using",
        "var",
        "void",
        "while",
        "with",
        "yield",
    }
)

PrismaIdentifierKind = Literal["model", "field", "relation"]


def is_prisma_identifier(name: str) -> bool:
    """Return whether ``name`` matches the Prisma identifier grammar."""
    return bool(PRISMA_IDENTIFIER_PATTERN.fullmatch(name))


def is_reserved_prisma_name(name: str) -> bool:
    """Return whether ``name`` is reserved for a Prisma model, enum, or type."""
    return (
        name.lower() in PRISMA_SCHEMA_KEYWORDS
        or name in PRISMA_SCALAR_TYPE_NAMES
        or name in PRISMA_CLIENT_RESERVED_NAMES
    )


def preferred_prisma_name(source: str) -> str:
    """Collapse a database name to a Prisma grammar candidate.

    Non-ASCII, punctuation, and whitespace become ``_``. Leading digits use
    the ``M_`` prefix. Reserved names take a trailing ``_`` so they do not
    collide with a source already named ``M_model``.
    """
    normalized = unicodedata.normalize("NFC", source)
    candidate = re.sub(r"[^A-Za-z0-9_]+", "_", normalized)
    candidate = re.sub(r"_+", "_", candidate).strip("_")
    if not candidate:
        candidate = "unnamed"
    if not candidate[0].isalpha():
        candidate = f"M_{candidate}"
    if is_reserved_prisma_name(candidate) or not is_prisma_identifier(candidate):
        candidate = f"{candidate}_"
    return candidate


def allocate_prisma_identifiers(
    requests: list[dict[str, str]],
    max_attempts: int = PRISMA_IDENTIFIER_MAX_ATTEMPTS,
) -> dict[str, Any]:
    """Allocate deterministic, collision-free Prisma identifiers.

    ``requests`` items use keys ``key``, ``kind``, ``namespace``, ``source``,
    and optional ``preferred``. Allocation order is source text, then
    namespace, then key, so input order cannot change the mapping.
    """
    names: dict[str, str] = {}
    mappings: list[dict[str, str]] = []
    used_by_namespace: dict[str, set[str]] = {}
    ordered = sorted(
        requests,
        key=lambda item: (item["source"], item["namespace"], item["key"]),
    )
    for request in ordered:
        used = used_by_namespace.setdefault(request["namespace"], set())
        base = request.get("preferred") or preferred_prisma_name(request["source"])
        candidate = base
        suffix = 2
        attempts = 0
        while (
            candidate in used
            or not is_prisma_identifier(candidate)
            or is_reserved_prisma_name(candidate)
        ):
            attempts += 1
            if attempts > max_attempts:
                return {
                    "ok": False,
                    "names": names,
                    "mappings": mappings,
                    "failure": {
                        "key": request["key"],
                        "kind": request["kind"],
                        "namespace": request["namespace"],
                        "source": request["source"],
                        "preferred": base,
                        "lastCandidate": candidate,
                        "attempts": attempts,
                        "maxAttempts": max_attempts,
                    },
                }
            candidate = f"{base}_{suffix}"
            suffix += 1
        used.add(candidate)
        names[request["key"]] = candidate
        mappings.append(
            {
                "kind": request["kind"],
                "namespace": request["namespace"],
                "source": request["source"],
                "generated": candidate,
            }
        )
    return {"ok": True, "names": names, "mappings": mappings}


def build_prisma_manifest(
    mappings: list[dict[str, str]], failure: dict[str, Any] | None = None
) -> dict[str, Any]:
    """Build the export manifest recorded beside a generated schema."""
    manifest: dict[str, Any] = {
        "contractVersion": PRISMA_IDENTIFIER_CONTRACT_VERSION,
        "mappings": mappings,
    }
    if failure is not None:
        manifest["failure"] = failure
    return manifest


def quote_prisma_string(value: str) -> str:
    """Quote a source identifier for ``@map`` / ``@@map``."""
    return json.dumps(value)
